File: benchmarks/node/bench.ts
import Redis from 'ioredis';
import { SlidingWindowLimiter } from '../../node/src/index.js';

const ITERATIONS = 10_000;
const CONCURRENCY = 100;
const redis = new Redis();
const limiter = new SlidingWindowLimiter(redis, { limit: ITERATIONS * 2, window: 60 });

async function run() {
  console.log(`Benchmarking SlidingWindowLimiter — ${ITERATIONS} requests, concurrency ${CONCURRENCY}\n`);

  // Sequential
  const seqStart = performance.now();
  for (let i = 0; i < ITERATIONS; i++) {
    await limiter.check(`bench:seq:${i % 100}`);
  }
  const seqMs = performance.now() - seqStart;

  // Concurrent batches
  const conStart = performance.now();
  const batches = Math.ceil(ITERATIONS / CONCURRENCY);
  for (let b = 0; b < batches; b++) {
    const batch = Array.from({ length: CONCURRENCY }, (_, i) =>
      limiter.check(`bench:con:${(b * CONCURRENCY + i) % 100}`),
    );
    await Promise.all(batch);
  }
  const conMs = performance.now() - conStart;

  console.log(`Sequential  : ${seqMs.toFixed(1)}ms total | ${(seqMs / ITERATIONS).toFixed(3)}ms/op | ${Math.round(ITERATIONS / (seqMs / 1000))} ops/s`);
  console.log(`Concurrent  : ${conMs.toFixed(1)}ms total | ${(conMs / ITERATIONS).toFixed(3)}ms/op | ${Math.round(ITERATIONS / (conMs / 1000))} ops/s`);

  await redis.quit();
}

run().catch(console.error);

File: benchmarks/python/bench.py
#!/usr/bin/env python3
"""Benchmark SlidingWindowLimiter against a local Redis instance."""
import time
import redis as redis_lib
from rate_limiter import SlidingWindowLimiter

ITERATIONS = 10_000

r = redis_lib.Redis()
limiter = SlidingWindowLimiter(r, limit=ITERATIONS * 2, window=60)

print(f"Benchmarking SlidingWindowLimiter — {ITERATIONS} sequential requests\n")

start = time.perf_counter()
for i in range(ITERATIONS):
    limiter.check(f"bench:{i % 100}")
elapsed = time.perf_counter() - start

per_op_ms = (elapsed / ITERATIONS) * 1000
ops_per_sec = int(ITERATIONS / elapsed)

print(f"Total   : {elapsed * 1000:.1f}ms")
print(f"Per op  : {per_op_ms:.3f}ms")
print(f"Ops/s   : {ops_per_sec}")
